### ISSUE TITLE
gcc/ld calls in the background leaving lots of temp files behind

### ISSUE THREAD

**Issue by jsievers** [Legrandin/pycryptodome#510]
We have narrowed down this to be new behavior introduced between version 3.9.9 and 3.10.1, most probably due to the recent ABI wheel changes:

on a system which is scheduling a lot of python processes using pycryptodome 3.10.1, we noticed lots of temp files (*.c, *.o, *.ld, *.res, *.le) were created (most of them 0 bytes size) slowly but steadily filling up the disk to the 100% inode usage limit.

it looks like for each python process, there are gcc and ld processes spawned.

forcing a downgrade to 3.9.9, the problem does not occur (no temp files created/left over).

In some of the temp files, we noticed messages like
```
/usr/bin/ld: cannot find -l/var/vcap/packages/monitoring-python-packages/Crypto/Util/../Util/_cpuid_c.cpython-36m-x86_64-linux-gnu.so
```

while there is no such file, there is an abi3 file
`/var/vcap/packages/monitoring-python-packages/Crypto/Util/_cpuid_c.abi3.so`

we are using python 3.6.12

```
$ uname -a
Linux 256dcd92-ae11-45a0-ae6d-077430b937a7 5.4.0-52-generic #57~18.04.1-Ubuntu SMP Thu Oct 15 14:04:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

$ gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1.3) 2.27
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

**Comment by kris7t** [comment]
I am affected by this on Arch Linux with pycryptodome 3.10.1. However, 3.10.0 seems to be working fine.

**Comment by Legrandin** [comment]
gcc/ld are (and should be) only called when building the library from source, possibly into a wheel.

Do you mean that gcc/ld are called (and all the various zero-size files are created) when *invoking* the library?

**Comment by kris7t** [comment]
I was getting the `ld` calls when trying to invoke `youtube-dl`, which depends on this library. So I presume it was at invoke time, unless `youtuble-dl` is rebuilding the library each time it is called.

**Comment by serge-name** [comment]
I have noticed the same. I use Python3 on Debian 10 (package version `3.7.3-2+deb10u3`) and `pycryptodomex==3.10.1`

`strace` shows that a lot of gcc executions happen in runtime, like this:

```
578152 12:06:53.029944 execve("/usr/bin/gcc", ["/usr/bin/gcc", "-Wl,-t", "-o", "/tmp/tmpycjhoim8", "-l/edx/app/sso/venv/lib/python3.7/site-packages/Cryptodome/Util/../Cipher/_raw_ecb.cpython-37m-x86_64-linux-gnu.so"], 0x20bfe10 /* 99 vars */ <unfinished ...>
578157 12:06:53.132358 execve("/usr/bin/gcc", ["/usr/bin/gcc", "-Wl,-t", "-o", "/tmp/tmp0cpmldz6", "-l…/site-packages/Cryptodome/Util/../Cipher/_raw_cbc.cpython-37m-x86_64-linux-gnu.so"], 0x20bfe10 /* 99 vars */ <unfinished ...>
578162 12:06:53.267555 execve("/usr/bin/gcc", ["/usr/bin/gcc", "-Wl,-t", "-o", "/tmp/tmpmjhr6mow", "-l…/site-packages/Cryptodome/Util/../Cipher/_raw_cfb.cpython-37m-x86_64-linux-gnu.so"], 0x20bfe10 /* 99 vars */ <unfinished ...>
578167 12:06:53.407119 execve("/usr/bin/gcc", ["/usr/bin/gcc", "-Wl,-t", "-o", "/tmp/tmpzu0m0ph_", "-l…/site-packages/Cryptodome/Util/../Cipher/_raw_ofb.cpython-37m-x86_64-linux-gnu.so"], 0x20f2190 /* 99 vars */ <unfinished ...>
578172 12:06:53.535894 execve("/usr/bin/gcc", ["/usr/bin/gcc", "-Wl,-t", "-o", "/tmp/tmp3dj_osy1", "-l…/site-packages/Cryptodome/Util/../Cipher/_raw_ctr.cpython-37m-x86_64-linux-gnu.so"], 0x20f2190 /* 99 vars */ <unfinished ...>
```

There are no `*.cpython-37m-x86_64-linux-gnu.so` libraries, only `*.abi3.so`. Exactly as in the issue.

I have made several experiments on a clear venv and found some interesting facts:

1) `pip install pycryptodomex==3.10.1`: no extra gcc executions

2) `pip install pycryptodomex==3.10.1 cffi==1.14.6`: this case gives us the same issue

After that I tried to build `pycryptodomex-3.10.*.whl` from sources:

1) `v3.10.0` produces `pycryptodomex-3.10.0-cp37-cp37m-linux_x86_64.whl`containing `*.cpython-37m-x86_64-linux-gnu.so`, no extra gcc executions

2) `v3.10.1` produces `pycryptodomex-3.10.1-cp35-abi3-linux_x86_64.whl` containing `*.cpython-37m-x86_64-linux-gnu.so` (not `*.abi3.so`! And I have no idea why.), no extra gcc executions.

Bottomline:

If there is `cffi` package exists in a virtualenv, and no `*.cpython-37m-x86_64-linux-gnu.so` libraries exist, we get the issue.

**Comment by Legrandin** [comment]
Thanks @serge-name, that is very helpful. I still can't reproduce it, but I am not yet using that combination of Debian and Python you mention.

I can at least see that `gcc` is (surprisingly to me) called: by the function ` _findLib_gcc` in `ctypes` (https://github.com/python/cpython/blob/main/Lib/ctypes/util.py#L102), which is invoked by `ctypes.util.find_library()`, which is called by `cffi` (https://foss.heptapod.net/pypy/cffi/-/blob/branch/default/cffi/api.py#L818) after we call `ffi.dlopen`.

` _findLib_gcc` does indeed create a temporary file, but it seems correctly closed (and erased) at least on Python 3.9.

I will keep looking into this.

**Comment by Legrandin** [comment]
Fixed in version 3.10.3.
